File: Python/Subsets II.py
"""
Given a collection of integers that might contain duplicates, S, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If S = [1,2,2], a solution is:

[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
"""

class Solution:
    # @param num, a list of integer
    # @return a list of lists of integer
    def subsetsWithDup(self, S):
        result = []
        self.subsetsWithDupRecu(result, [], sorted(S))
        return result
    
    def subsetsWithDupRecu(self, result, cur, S):
        if len(S) == 0  and cur not in result:
            result.append(cur)
        elif S:
            self.subsetsWithDupRecu(result, cur, S[1:])
            self.subsetsWithDupRecu(result, cur + [S[0]], S[1:])

class Solution:
    # @param num, a list of integer
    # @return a list of lists of integer
    def subsetsWithDup(self, S):
        if len(S) == 0:
            return [[]]
        num = 2 ** len(S)
        result = []
        for i in range(num):
            tmp = []
            idx = 0
            j = i
            while j > 0:
                if j & 1:
                    tmp.append(S[idx])
                idx +=1
                j >>=1
            tmp = sorted(tmp)
            if tmp not in result:
                result.append(tmp)
        return result

File: Java/Subsets II.java
/*
Given a collection of integers that might contain duplicates, nums, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If nums = [1,2,2], a solution is:

[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
*/

public class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);
        generateSubset(nums, result, new ArrayList<>(), 0);
        return result;
    }
    
    private void generateSubset(int[] nums, List<List<Integer>> result, List<Integer> cur, int step) {
        result.add(new ArrayList<>(cur));
        
        for(int i = step ; i < nums.length ; i ++) {
            if(i != step && nums[i-1] == nums[i]) continue;
            cur.add(nums[i]);
            generateSubset(nums, result, cur, i + 1);
            cur.remove(cur.size() - 1);
        }
    }
}

public class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);
        int count = (1 << nums.length) - 1;
        
        while(count >= 0){
            List<Integer> cur = new ArrayList<>();
            int idx = 0;
            while(idx < nums.length) {
                if(((count >> idx) & 1) == 1) {
                    cur.add(nums[idx]);
                }
                idx ++;
            }
            if(!result.contains(cur)) {
                result.add(cur);
            }
            count --;
        }
        return result;
    }
}
